Proguard-related crash on Fragment recreation

Reproduced with MvRx v1.1.0 and the TodoMvRx sample app.

#### Steps to Reproduce
1. Developer Options -> Don't Keep Activities
2. Launch minified "TodoMvRx" sample app _(note that debugMode in the viewmodel must be false)_
3. Navigate away from TodoMvRx app in recent apps list
4. Return to TodoMvRx app from the recent apps list

#### To Fix the Crash
Add the following line to `proguard-rules.pro`
`
-keep class kotlin.Metadata { *; }
`


#### Crash Log
`
2019-10-02 15:20:35.934 26866-26866/com.airbnb.mvrx.todoapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.airbnb.mvrx.todoapp, PID: 26866
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.airbnb.mvrx.todoapp/com.airbnb.mvrx.todomvrx.MainActivity}: java.lang.IllegalStateException: TasksState has no primary constructor!
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3270)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalStateException: TasksState has no primary constructor!
        at com.airbnb.mvrx.PersistStateKt.a(PersistState.kt:97)
        at com.airbnb.mvrx.MvRxViewModelStore$restoreViewModel$1.a(MvRxViewModelStore.kt:124)
        at com.airbnb.mvrx.MvRxViewModelStore$restoreViewModel$1.a(MvRxViewModelStore.kt:20)
        at com.airbnb.mvrx.RealMvRxStateFactory.a(MvRxStateFactory.kt:27)
        at com.airbnb.mvrx.MvRxViewModelProvider.a(MvRxViewModelProvider.kt:49)
        at com.airbnb.mvrx.MvRxViewModelProvider.a(MvRxViewModelProvider.kt:47)
        at com.airbnb.mvrx.MvRxViewModelStore.a(MvRxViewModelStore.kt:124)
        at com.airbnb.mvrx.MvRxViewModelStore.a(MvRxViewModelStore.kt:109)
        at com.airbnb.mvrx.MvRxViewModelStore.a(MvRxViewModelStore.kt:84)
        at com.airbnb.mvrx.MvRxViewModelStore.a(MvRxViewModelStore.kt:74)
        at com.airbnb.mvrx.BaseMvRxActivity.onCreate(BaseMvRxActivity.kt:33)
        at com.airbnb.mvrx.todomvrx.MainActivity.onCreate(MainActivity.kt:25)
        at android.app.Activity.performCreate(Activity.java:7802)
        at android.app.Activity.performCreate(Activity.java:7791)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1306)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3245)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
`

--- a/build.gradle
+++ b/build.gradle
@@ -24,6 +24,13 @@ buildscript {
         classpath 'net.sf.proguard:proguard-gradle:6.1.1'
         classpath 'com.vanniktech:gradle-maven-publish-plugin:0.8.0'
     }
+
+    configurations.all {
+        resolutionStrategy {
+            // Use Proguard 6.2.0
+            force 'net.sf.proguard:proguard-gradle:6.2.0'
+        }
+    }
 }
 
 allprojects {

--- a/gradle.properties
+++ b/gradle.properties
@@ -14,3 +14,5 @@ POM_DEVELOPER_NAME=Airbnb
 POM_INCEPTION_YEAR=2018
 android.useAndroidX = true
 android.enableJetifier = true
+# Set to false if you want to use Proguard
+android.enableR8=true

--- a/sample/proguard-project.pro
+++ b/sample/proguard-project.pro
@@ -124,27 +124,26 @@
     <init>(...);
     <fields>;
 }
-# Proguard bug: https://sourceforge.net/p/proguard/bugs/731/
-#-if @com.squareup.moshi.JsonClass class **$*$*
-#-keep class <1>_<2>_<3>JsonAdapter {
-#    <init>(...);
-#    <fields>;
-#}
-#-if @com.squareup.moshi.JsonClass class **$*$*$*
-#-keep class <1>_<2>_<3>_<4>JsonAdapter {
-#    <init>(...);
-#    <fields>;
-#}
-#-if @com.squareup.moshi.JsonClass class **$*$*$*$*
-#-keep class <1>_<2>_<3>_<4>_<5>JsonAdapter {
-#    <init>(...);
-#    <fields>;
-#}
-#-if @com.squareup.moshi.JsonClass class **$*$*$*$*$*
-#-keep class <1>_<2>_<3>_<4>_<5>_<6>JsonAdapter {
-#    <init>(...);
-#    <fields>;
-#}
+-if @com.squareup.moshi.JsonClass class **$*$*
+-keep class <1>_<2>_<3>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*
+-keep class <1>_<2>_<3>_<4>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*$*
+-keep class <1>_<2>_<3>_<4>_<5>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*$*$*
+-keep class <1>_<2>_<3>_<4>_<5>_<6>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
 
 
 # 

--- a/todomvrx/proguard-rules.pro
+++ b/todomvrx/proguard-rules.pro
@@ -1,21 +1,49 @@
-# Add project specific ProGuard rules here.
-# You can control the set of applied configuration files using the
-# proguardFiles setting in build.gradle.
+# This file contains project specific proguard rules
+
+-android
+
+# Print out the full proguard config used for every build
+-printconfiguration build/outputs/fullProguardConfig.pro
+
 #
-# For more details, see
-#   http://developer.android.com/guide/developing/tools/proguard.html
+# Keep rules that are used because MvRx uses Kotlin, Kotlin reflection and RxJava. These are not defined in the
+# lib as they are not specific to the lib itself but need to be most likely present in any project that uses
+# Kotlin, Kotlin reflection and RxJava.
+#
 
-# If your project uses WebView with JS, uncomment the following
-# and specify the fully qualified class name to the JavaScript interface
-# class:
-#-keepclassmembers class fqcn.of.javascript.interface.for.webview {
-#   public *;
-#}
+# These classes are used via kotlin reflection and the keep might not be required anymore once Proguard supports
+# Kotlin reflection directly.
+-keep interface kotlin.reflect.jvm.internal.impl.builtins.BuiltInsLoader
+-keep class kotlin.reflect.jvm.internal.impl.builtins.BuiltInsLoaderImpl
+-keep class kotlin.reflect.jvm.internal.impl.load.java.FieldOverridabilityCondition
+-keep class kotlin.reflect.jvm.internal.impl.load.java.ErasedOverridabilityCondition
+-keep class kotlin.reflect.jvm.internal.impl.load.java.JavaIncompatibilityRulesOverridabilityCondition
 
-# Uncomment this to preserve the line number information for
-# debugging stack traces.
-#-keepattributes SourceFile,LineNumberTable
+# If Companion objects are instantiated via Kotlin reflection and they extend/implement a class that Proguard
+# would have removed or inlined we run into trouble as the inheritance is still in the Metadata annotation
+# read by Kotlin reflection.
+# FIXME Remove if Kotlin reflection is supported by Pro/Dexguard
+-if class **$Companion extends **
+-keep class <2>
+-if class **$Companion implements **
+-keep class <2>
 
-# If you keep the line number information, uncomment this to
-# hide the original source file name.
-#-renamesourcefileattribute SourceFile
+# https://medium.com/@AthorNZ/kotlin-metadata-jackson-and-proguard-f64f51e5ed32
+-keep class kotlin.Metadata { *; }
+
+# https://stackoverflow.com/questions/33547643/how-to-use-kotlin-with-proguard
+-dontwarn kotlin.**
+
+# RxJava
+-keepclassmembers class rx.internal.util.unsafe.*ArrayQueue*Field* {
+   long producerIndex;
+   long consumerIndex;
+}
+
+#
+# Keep rules that are unique to the sample project because it uses epoxy
+#
+
+# The code is safeguarded against API28 use so we can just ignore the warning bout Handler.createAsync(looper) not existing.
+# Can be removed when compiled against API >=28
+-dontwarn com.airbnb.epoxy.EpoxyAsyncUtil

